### PR TITLE
feat: Add Sentry.start

### LINKED ISSUE #432: Calling initializer to cause global side-effect is not very swifty
I just had time to have a look at the 5.0 beta. Overall I like the simplification of the API, but having the initialization of the shared client instance happen through a call to an initializer looks very weird to me. I would expect most or all initializers to be free of side-effects. Seeing the result of an initializer be discarded immediately looks very much non-idiomatic.

Since the side-effect is the only thing this operation is actually doing (as opposed to initializing and returning an instance to be used by the caller), a static function would be much more appropriate here.

So effectively, I'm proposing this:

```
 _ = SentrySDK(options: [
    "dsn": "___PUBLIC_DSN___",
    "debug": true
])
```

should instead turn into this:

```
SentrySDK.initialize(options: [
    "dsn": "___PUBLIC_DSN___",
    "debug": true
])
```

(I'm not really concerned about the method name, setUp or install could work just as well)

### PR DESCRIPTION
**Pull request by HazAT**
Fixes #432

Add `Sentry.start` function for Swift which is a static call while keeping the old functionality.
We will update docs so it reflects that we prefer this.

Other Sentry SDKs have `Sentry.init`, `init` is a reserved word in both Objective-C and Swift therefore we can't use it. We tried `Sentry.initialize` but still, could be confused with [`NSObject.initialize()`](https://developer.apple.com/documentation/objectivec/nsobject/1418639-initialize).

Using the standard initializer as we do right now is misleading since users think they need to take care of the instance (which they don't):

```swift
_ = Sentry(options: [ ... ]);
```

Currently, we are leaning towards:

```swift
Sentry.start(options: [ ... ]);
```

Other options to consider:

```swift

Sentry.with(options: [ ... ]);
Sentry.initialize(options: [ ... ]);
Sentry.configure(options: [ ... ]);

```

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -37,7 +37,7 @@ import Sentry
 
 // ....
 
-[SentrySDK initWithOptions:@{
+[SentrySDK initializeWithOptions:@{
```
Comment (philipphofmann) [inline README.md:40]: I think this should be `withOptions`
```suggestion
[SentrySDK withOptions:@{
```

File: README.md [hunk 2]
```diff
@@ -26,7 +26,7 @@ import Sentry
 
 // ....
 
- _ = SentrySDK(options: [
+SentrySDK.initialize(options: [
```
Comment (philipphofmann) [inline README.md:29]: I think this should be
```suggestion
SentrySDK.with(options: [
```

File: CHANGELOG.md [hunk 3]
```diff
@@ -2,6 +2,8 @@
 
 ## Unreleased
 
+- feat: Add Sentry initialize #441
```
Comment (philipphofmann) [inline CHANGELOG.md:5]: Maybe we should update the info here as well.

File: Sources/Sentry/include/SentrySDK.h [hunk 4]
```diff
@@ -34,19 +34,25 @@ SENTRY_NO_INIT
...
+*/
 + (instancetype)initWithOptionsObject:(SentryOptions *)options NS_SWIFT_NAME(init(options:));
 
 /**
- starts sentry with options and starts crash handler
- 
- Inits and configures Sentry (SentryHub, SentryClient) and starts crash handler
+ * Use [SentrySDK initializeWithOptions]
  */
 + (instancetype)initWithOptions:(NSDictionary<NSString *, id> *)optionsDict NS_SWIFT_NAME(init(options:));
```
Comment (philipphofmann) [inline Sources/Sentry/include/SentrySDK.h:44]: Shouldn't we mark this method and `initWithOptionsObject` as deprecated?

**Comment by philipphofmann** [comment]
I also agree with @marandaneto that `with()` is maybe not the best option, since in other languages it has a different concept. For me `with` stands more for a builder pattern: `Sentry.with(foo).with(bar).start()`. @marandaneto `initialize` is kind of out, because it overlaps with [`NSObject.initialize()`](https://developer.apple.com/documentation/objectivec/nsobject/1418639-initialize).

For me `start()` makes sense as it matches with `Sentry.close()`, that we already have on [Android](https://github.com/getsentry/sentry-android/blob/3141cb8ac16c52788832f21d2486945af2d5daa6/sentry-core/src/main/java/io/sentry/core/Sentry.java#L150) and [Dotnet](https://github.com/getsentry/sentry-dotnet/blob/master/src/Sentry/SentrySdk.cs#L116).

Just for the record Google Firebase uses [`FirebaseApp.configure()`](https://firebase.google.com/docs/ios/setup#initialize-firebase) and Microsoft App Center uses [`MSAppCenter.start(...)`](https://docs.microsoft.com/en-us/appcenter/sdk/getting-started/ios#4-start-the-sdk).

**Comment by marandaneto** [comment]
> I also agree with @marandaneto that `with()` is maybe not the best option, since in other languages it has a different concept. For me `with` stands more for a builder pattern: `Sentry.with(foo).with(bar).start()`. @marandaneto `initialize` is kind of out, because it overlaps with [`NSObject.initialize()`](https://developer.apple.com/documentation/objectivec/nsobject/1418639-initialize).
> 
> For me `start()` makes sense as it matches with `Sentry.close()`, that we already have on [Android](https://github.com/getsentry/sentry-android/blob/3141cb8ac16c52788832f21d2486945af2d5daa6/sentry-core/src/main/java/io/sentry/core/Sentry.java#L150) and [Dotnet](https://github.com/getsentry/sentry-dotnet/blob/master/src/Sentry/SentrySdk.cs#L116).
> 
> Just for the record Google Firebase uses [`FirebaseApp.configure()`](https://firebase.google.com/docs/ios/setup#initialize-firebase) and Microsoft App Center uses [`MSAppCenter.start(...)`](https://docs.microsoft.com/en-us/appcenter/sdk/getting-started/ios#4-start-the-sdk).

https://firebase.google.com/docs/crashlytics/customize-crash-reports?platform=ios#objective-c_4
it seems that if you only Fabric Crashlytisc they actually use `with`, see `[Fabric with:@[[Crashlytics class]]];`

**Comment by bruno-garcia** [comment]
> I also agree with @marandaneto that with() is maybe not the best option, since in other languages it has a different concept. 

In all other languages we get away with `init`. So this change is to write something that makes sense in the swift ecosystem mainly. @hebertialmeida suggested `with` and in my (non swift dev) opinion, `Sentry SDK with options` does make sense. and `Sentry SDK close` goes well with it.

I wouldn't say `start/close` goes that much together since with `start` I'd expect `close` and with `open` I expect `close` even. But I believe we're trying to find a **idiomatic** solution here, in this case I'd suggest to ask a few swift devs and pick the most voted approach. What do you think?

File: CHANGELOG.md (hunk 3 above)
Comment (bruno-garcia) [inline CHANGELOG.md:5]: Good point.
```suggestion
- feat: Add Sentry initialization function 'with' #441
```

File: Sources/Sentry/SentrySDK.m [hunk 5]
```diff
@@ -42,24 +42,32 @@ + (void)setCurrentHub:(SentryHub *)hub {
 }
  
 + (id)initWithOptions:(NSDictionary<NSString *,id> *)optionsDict {
+    [SentrySDK withOptions:optionsDict];
+    return nil;
+}
+
++ (id)initWithOptionsObject:(SentryOptions *)options {
```
Comment (bruno-garcia) [inline Sources/Sentry/SentrySDK.m:49]: Does this show up as an overload in swift?

It was suggested the ideal API would be

```swift
SentrySDK.with(options: .init(
    dsn: "https://...”,
    debug: true,
    logLevel: "verbose",
    enableAutoSessionTracking: true,
    sessionTrackingIntervalMillis: 5000
))
```

As opposed to the Dictionary approach (which makes total sense for Objective-C

**Comment by ahti** [comment]
Fwiw, the [Swift API design guidelines](https://swift.org/documentation/api-design-guidelines/#strive-for-fluent-usage) state "Name functions and methods according to their side-effects [...] Those with side-effects should read as imperative verb phrases, e.g., print(x), x.sort(), x.append(y)."

With that in mind (and the fact that `initialize` is taken by `NSObject`) I think from the options listed above `configure` would be the most clear one.

**Comment by bruno-garcia** [comment]
> Fwiw, the Swift API design guidelines state "Name functions and methods according to their side-effects [...] Those with side-effects should read as imperative verb phrases, e.g., print(x), x.sort(), x.append(y)."

In that case `configure` wouldn't be the best option since Sentry SDK's `init` function not only configures, but has other side effects like starting a background worker, starting session tracking and other integrations. 

So `start` would be best using your argument as it _starts the SDK with these options_.

File: Sources/Sentry/SentrySDK.m [hunk 6]
```diff
@@ -42,24 +42,32 @@ + (void)setCurrentHub:(SentryHub *)hub {
 }
  
 + (id)initWithOptions:(NSDictionary<NSString *,id> *)optionsDict {
```
Comment (marandaneto) [inline Sources/Sentry/SentrySDK.m:44]:
```suggestion
+ (id)startWithOptions:(NSDictionary<NSString *,id> *)optionsDict {
```

should this be renamed too?

File: Sources/Sentry/SentrySDK.m [hunk 7]
```diff
@@ -42,24 +42,32 @@ + (void)setCurrentHub:(SentryHub *)hub {
...
+    [SentrySDK startWithOptions:optionsDict];
+    return nil;
+}
+
++ (id)initWithOptionsObject:(SentryOptions *)options {
```
Comment (marandaneto) [inline Sources/Sentry/SentrySDK.m:49]:
```suggestion
+ (id)startWithOptionsObject:(SentryOptions *)options {
```

also here?

File: Sources/Sentry/SentrySDK.m (hunk 6 above)
Comment (marandaneto) [reply Sources/Sentry/SentrySDK.m:44]: if it's the deprecated method, I'd add a "javadoc"

File: Sources/Sentry/SentrySDK.m (hunk 7 above)
Comment (marandaneto) [reply Sources/Sentry/SentrySDK.m:49]: if it's the deprecated method, I'd add a "javadoc"

File: Sources/Sentry/SentrySDK.m (hunk 6 above)
Comment (HazAT) [reply Sources/Sentry/SentrySDK.m:44]: These are the old `init`, if we remove them we have a breaking change.

Comment (marandaneto) [reply Sources/Sentry/SentrySDK.m:44]: @HazAT yep but whats about a deprecated javadoc? like you did with other methods eg `initWithOptionsObject`

File: Sources/Sentry/SentrySDK.m [hunk 8]
```diff
@@ -42,24 +42,32 @@ + (void)setCurrentHub:(SentryHub *)hub {
 }
  
 + (id)initWithOptions:(NSDictionary<NSString *,id> *)optionsDict {
+    [SentrySDK startWithOptions:optionsDict];
```
Comment (philipphofmann) [inline Sources/Sentry/SentrySDK.m:45]: Can we mark the old init functions as deprecated?

Comment (HazAT) [reply Sources/Sentry/SentrySDK.m:45]: They are both marked as deprecated
https://github.com/getsentry/sentry-cocoa/pull/441/files#diff-e98d4373fbfc44d86abfa859f8883ee3R36-R45

Comment (philipphofmann) [reply Sources/Sentry/SentrySDK.m:45]: Ah sorry, didn't see it. Thanks for pointing it out.

**Pull request merged**